2_process.smk

```diff
@@ -0,0 +1,155 @@
+    :param summary_file: csv file with how many sequences are in each lake
+
+    """
+    sequence_counts = []
+    for sequences_file in lake_sequence_files_input:
+        num_sequences = np.load(sequences_file).shape[0] 
+        sequence_counts.append(num_sequences)
+    df_counts = pd.DataFrame(data={
+        'sequences_file': lake_sequence_files_input,
+        'num_sequences': sequence_counts
+    })
+    df_counts.to_csv(summary_file, index=False)
+
+
+def dynamic_filenames(site_id, file_category = 'dynamic_mntoha'):
+    """
+    Return the files that contain dynamic data that are needed to construct
+    sequences for a given lake.
+
+    This function also triggers four checkpoints:
+    1. fetch_mntoha_metadata to get lake_metadata.csv
+    2. unzip_archive for this lake's drivers
+    3. unzip_archive for this lake's clarity
+    3. unzip_archive for this lake's ice flags
+
+    :param site_id: NHDHR lake ID
+    :param file_category: category of files, e.g., dynamic_mntoha. Used by
+    unzip_archive to determine parent directory.
+    :returns: List of 3 dynamic filenames: drivers, clarity, and ice flags
+
+    """
+    # make this function depend on fetch_mntoha_metadata
+    # needed because lake_metadata.csv is used to determine dynamic files
+    lake_metadata_file = checkpoints.fetch_mntoha_metadata.get().output[0]
+    lake_metadata = pd.read_csv(lake_metadata_file)
+    lake = lake_metadata.loc[lake_metadata['site_id']==site_id].iloc[0]
+    # also make this function depend on unzip_archive
+    # needed to link unzipped files with unzip_archive rule
+    drivers_directory = f'inputs_{lake.group_id}'
+    unzip_archive_drivers = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=drivers_directory
+    ).output[0]
+    clarity_directory = f'clarity_{lake.group_id}'
+    unzip_archive_clarity = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=clarity_directory
+    ).output[0]
+    ice_flags_directory = f'ice_flags_{lake.group_id}'
+    unzip_archive_ice_flags = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=ice_flags_directory
+    ).output[0]
+
+    # dynamic filenames
+    drivers_file = f'{unzip_archive_drivers}/{lake.meteo_filename}'
+    clarity_file = f'{unzip_archive_clarity}/gam_{lake.site_id}_clarity.csv'
+    ice_flags_file = f'{unzip_archive_ice_flags}/pb0_{lake.site_id}_ice_flags.csv'
+    return [drivers_file, clarity_file, ice_flags_file]
+
+
+# Summarize training sequences
+rule process_mntoha:
+    input:
+        mntoha_lake_sequence_files
+    output:
+        "2_process/out/mntoha_sequences/mntoha_sequences_summary.csv"
+    run:
+        save_sequences_summary(input, output[0])
+
+
+# Create .npy of input/output sequences for one lake to use for training and testing
+rule mntoha_lake_sequences:
+    input:
```

Comment:
I'm still in the midst of reviewing `assemble_lake_data()`, but I see that while the drivers, clarity, and ice flag files returned by `dynamic_filenames()` are site-specific (since `snakemake` is mapping over the wildcard `site_id` to retrieve those files, the other two inputs, `"2_process/tmp/mntoha/lake_metadata_augmented.csv"` and `"2_process/tmp/mntoha/temperature_observations_interpolated.csv"` are not site-specific. Instead, those two dataframes get filtered to the appropriate site within `assemble_lake_data()`.

Two factors to consider re: the validity of my comment here: 1) I'm not fully up to speed on `snakemake` dependency tracking or what options would be on the table, and 2) this may be less relevant for a global model pipeline than it is for a site-specific model pipeline.

My thoughts: in a `targets` modeling workflow, we'd typically avoid passing in the _complete_ set of site metadata and _complete_ set of observations, because a change in the metadata or # of observations for _any_ site would then trigger a re-build of the sequences file file for _every_ site. Ditto if you added a new site. Instead I think we'd create an intermediary target for the metadata and another for the observations that map over the site_ids to generate site-specific objects/files that would then be passed into `assemble_lake_data()` alongside the other 3 files. I'm not certain what the equivalent approach would be in `snakemake` - maybe you would make a function or functions that would make use of the `site_id` wildcard to filter the observations/metadata to that site, save to a csv for that site, and return that site-specific csv? And then you could list that function(s) under inputs, here?

Reply:
Yes, that would speed up the pipeline whenever metadata or observations changed. I think you're right that this would make a bigger difference for training many site-specific models - adding new sites wouldn't require that the old sites get re-run.

Snakemake may or may not support that kind of workflow, but so far I haven't learned how to implement that in Snakemake. I think (big caveat: I'm still learning) that Snakemake looks at a file's modification date, and not its contents. So, say that our pipeline separates `lake_metadata.csv` into many site-specific files. Then, say we add a new lake to `lake_metadata.csv`. The function that breaks the csv apart into site-specific files would still overwrite all those files, even though it actually only needed to make one new file. Then, Snakemake would re-execute every downstream task.

So, I think it would take a lot of manual work outside of Snakemake to set up a pipeline in which new files that need to be remade are remade, and files that don't need to be remade are not.

Something like: 
1. Read in every existing file
2. Check to see if the new content that would be written to the file match the old content
3. Only write that new content to the file if they don't match.

Reply:
Ah okay, that's helpful context, and good to be aware of the constraints with snakemake. Definitely does not seem worth exploring at this point. Thanks!